[BUG] `npm publish` and `npm install` fail depending on `.npmrc` syntax used

### Current Behavior:

I am attempting to configure a project to install dependencies from NPM. I will be publishing the project to GitHub Packages as a private package. If I use this syntax in my project's `.npmrc`:

```
@my-org:registry=https://npm.pkg.github.com/
```

I can install dependencies from NPM using `npm install` on my local machine. However, I cannot publish to GitHub Packages using `npm publish`. NPM informs me that I'm not authenticated. If I use this syntax in my project's `.npmrc`:

```
registry=https://npm.pkg.github.com/my-org/
```

I can publish using `npm publish`, but I cannot install dependencies with `npm install`. NPM informs me that it's trying to install dependencies from GitHub Packages, rather than NPM.

### Expected Behavior:

Based on my reading, both syntaxes should be compatible with `npm install` and `npm publish`. However, it appears I can only use one or the other, based on my intended use.

### Steps To Reproduce:

1.  Install Node `v15.7.0` and NPM `7.4.3` via [nvm](https://github.com/nvm-sh/nvm).
2.  Log in to GitHub Packages with the command:

    ```shell
    npm login --scope=@my-org --registry=https://npm.pkg.github.com
    ```

3.  Check our `~/.npmrc` file in our home folder. It should read:

    ```
    @my-org:registry=https://npm.pkg.github.com/
    //npm.pkg.github.com/:_authToken=<auth-token-used-for-login>
    ```

4.  Create project with the following `package.json`:

    ```json
    {
      "name": "@my-org/my-package",
      "description": "A test.",
      "version": "1.0.0",
      "scripts": {
        "test": "echo \"Error: no test specified\" && exit 1"
      },
      "repository": {
        "type": "git",
        "url": "https://github.com/my-org/my-package.git"
      },
      "keywords": ["example"],
      "author": "Me",
      "license": "ISC",
      "bugs": {
        "url": "https://github.com/my-org/my-package/issues"
      },
      "homepage": "https://github.com/my-org/my-package",
      "dependencies": {
        "bootstrap": "^4.5.2"
      }
    }
    ```

5.  Add the following `.npmrc` to our project:

    ```
    @my-org:registry=https://npm.pkg.github.com/
    ```

6.  Run `npm install`. Installation should succeed.
7.  Run `npm publish`. Receive the following error:

    ```shell
    npm ERR! code ENEEDAUTH
    npm ERR! need auth This command requires you to be logged in.
    npm ERR! need auth You need to authorize this machine using `npm adduser`

    npm ERR! A complete log of this run can be found in:
    npm ERR!     /Users/my-user/.npm/_logs/2021-01-28T20_19_55_974Z-debug.log
    ```

8.  Change the project `.npmrc` to:

    ```
    registry=https://npm.pkg.github.com/my-org/
    ```

9.  Run `npm publish`. Publishing should succeed.
10. `rm -rf node_modules/ package-lock.json` in project.
11. Run `npm install`. Receive following error:

    ```
    npm ERR! code E401
    npm ERR! Incorrect or missing password.
    npm ERR! If you were trying to login, change your password, create an
    npm ERR! authentication token or enable two-factor authentication then
    npm ERR! that means you likely typed your password in incorrectly.
    npm ERR! Please try again, or recover your password at:
    npm ERR!     https://www.npmjs.com/forgot
    npm ERR!
    npm ERR! If you were doing some other operation then your saved credentials are
    npm ERR! probably out of date. To correct this please try logging in again with:
    npm ERR!     npm login

    npm ERR! A complete log of this run can be found in:
    npm ERR!     /Users/my-user/.npm/_logs/2021-01-28T20_38_20_711Z-debug.log
    ```

### Environment:

- OS:
  - MacOS Catalina 10.15.7 (Intel Mac Mini)
  - MacOS Big Sur 11.1 (Apple Silicon MacBook AIr)
- Node: 15.7.0
- npm:
  - 7.4.3
  - 7.5.2

I'm seeing something similar.  npm v7 was just "released" and now all our `npm publish` calls are breaking to our locally scoped repository.  Previously, doing these two commands would generate a usable .npmrc file for our environment:
```
npm config set @privatescope:registry https://my.server.org/repository/npm-internal/
npm adduser --registry=https://my.server.org/repository/npm-internal/ --always-auth
```
This still generates an .npmrc file that looks identical between npm v6 and npm v7.
```
@privatescope:registry =https://my.server.org/repository/npm-internal/
//my.server.org/repository/npm-internal/:_authToken=NpmToken.9fb28705-7e5b-3ea9-894b-06f7d898886c
```
However, attempting to use it fails on v7:
```
npm ERR! code ENEEDAUTH
npm ERR! need auth This command requires you to be logged in.
npm ERR! need auth You need to authorize this machine using `npm adduser`

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\user\AppData\Local\npm-cache\_logs\2021-02-02T20_51_38_576Z-debug.log
```

I have this problem 
$ npm i
npm WARN read-shrinkwrap This version of npm is compatible with lockfileVersion@1, but package-lock.json was generated for lockfileVersion@2. I'll try to do my best with it!
npm ERR! code ENOTFOUND
npm ERR! errno ENOTFOUND
npm ERR! network request to https://registry.npmjs.org/acorn/-/acorn-7.3.1.tgz failed, reason: getaddrinfo ENOTFOUND proxy.admworld.com
npm ERR! network This is a problem related to network connectivity.
npm ERR! network In most cases you are behind a proxy or have bad network settings.
npm ERR! network
npm ERR! network If you are behind a proxy, please make sure that the
npm ERR! network 'proxy' config is set properly.  See: 'npm help config'

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\Shahab\AppData\Roaming\npm-cache\_logs\2021-02-04T08_53_02_399Z-debug.log

--- a/test/lib/publish.js
+++ b/test/lib/publish.js
@@ -8,11 +8,11 @@ const credentials = {
     email: 'me@example.com',
   },
   'https://scope.specific.registry': {
-    token: 'some.regsitry.token',
+    token: 'some.registry.token',
     alwaysAuth: false,
   },
   'https://some.registry': {
-    token: 'some.regsitry.token',
+    token: 'some.registry.token',
     alwaysAuth: false,
   },
   'https://registry.npmjs.org/': {
@@ -22,9 +22,6 @@ const credentials = {
 }
 const config = {
   list: [defaults],
-  getCredentialsByURI: (uri) => {
-    return credentials[uri]
-  },
 }
 
 const registryCredentials = (t, registry) => {
@@ -155,9 +152,10 @@ t.test('re-loads publishConfig if added during script process', (t) => {
   })
 })
 
-t.test('should not log if silent', (t) => {
+t.test('should not log if silent (dry run)', (t) => {
   t.plan(2)
 
+  const registry = 'https://registry.npmjs.org'
   const testDir = t.testdir({
     'package.json': JSON.stringify({
       name: 'my-cool-pkg',
@@ -171,9 +169,14 @@ t.test('should not log if silent', (t) => {
         json: false,
         defaultTag: 'latest',
         dryRun: true,
-        registry: 'https://registry.npmjs.org/',
+        registry,
       },
-      config,
+      config: {
+        ...config,
+        getCredentialsByURI: () => {
+          throw new Error('should not call getCredentialsByURI in dry run')
+        },
+      },
     },
     '../../lib/utils/tar.js': {
       getContents: () => ({}),
@@ -195,7 +198,7 @@ t.test('should not log if silent', (t) => {
     libnpmpack: async () => '',
     libnpmpublish: {
       publish: (manifest, tarData, opts) => {
-        throw new Error('should not call libnpmpublish!')
+        throw new Error('should not call libnpmpublish in dry run')
       },
     },
   })
@@ -207,8 +210,10 @@ t.test('should not log if silent', (t) => {
   })
 })
 
-t.test('should log tarball contents', (t) => {
+t.test('should log tarball contents (dry run)', (t) => {
   t.plan(3)
+
+  const registry = 'https://registry.npmjs.org'
   const testDir = t.testdir({
     'package.json': JSON.stringify({
       name: 'my-cool-pkg',
@@ -222,12 +227,12 @@ t.test('should log tarball contents', (t) => {
         json: false,
         defaultTag: 'latest',
         dryRun: true,
-        registry: 'https://registry.npmjs.org/',
+        registry,
       },
       config: {
         ...config,
         getCredentialsByURI: () => {
-          throw new Error('should not call getCredentialsByURI!')
+          throw new Error('should not call getCredentialsByURI in dry run')
         }},
     },
     '../../lib/utils/tar.js': {
@@ -247,7 +252,7 @@ t.test('should log tarball contents', (t) => {
     libnpmpack: async () => '',
     libnpmpublish: {
       publish: () => {
-        throw new Error('should not call libnpmpublish!')
+        throw new Error('should not call libnpmpublish in dry run')
       },
     },
   })
@@ -277,12 +282,15 @@ t.test('shows usage with wrong set of arguments', (t) => {
 
 t.test('throws when invalid tag', (t) => {
   t.plan(1)
+
+  const registry = 'https://registry.npmjs.org'
+
   const publish = requireInject('../../lib/publish.js', {
     '../../lib/npm.js': {
       flatOptions: {
         json: false,
         defaultTag: '0.0.13',
-        registry: 'https://registry.npmjs.org/',
+        registry,
       },
       config,
     },
@@ -296,7 +304,9 @@ t.test('throws when invalid tag', (t) => {
 })
 
 t.test('can publish a tarball', t => {
-  t.plan(3)
+  t.plan(4)
+
+  const registry = 'https://registry.npmjs.org/'
   const testDir = t.testdir({
     package: {
       'package.json': JSON.stringify({
@@ -322,9 +332,12 @@ t.test('can publish a tarball', t => {
       flatOptions: {
         json: true,
         defaultTag: 'latest',
-        registry: 'https://registry.npmjs.org/',
+        registry,
       },
-      config,
+      config: {
+        ...config,
+        getCredentialsByURI: registryCredentials(t, registry),
+      },
     },
     '../../lib/utils/tar.js': {
       getContents: () => ({
@@ -355,14 +368,19 @@ t.test('can publish a tarball', t => {
 })
 
 t.test('throw if not logged in', async t => {
-  t.plan(1)
+  t.plan(2)
+  const registry = 'https://unauthed.registry'
+
   const publish = requireInject('../../lib/publish.js', {
     '../../lib/npm.js': {
       flatOptions: {
         json: false,
-        registry: 'https://unauthed.registry',
+        registry,
       },
-      config,
+      config: {
+        ...config,
+        getCredentialsByURI: registryCredentials(t, registry),
+      },
     },
   })
 
@@ -375,9 +393,10 @@ t.test('throw if not logged in', async t => {
 })
 
 t.test('read registry only from publishConfig', t => {
-  t.plan(3)
+  t.plan(4)
 
-  const publishConfig = { registry: 'https://some.registry' }
+  const registry = 'https://some.registry'
+  const publishConfig = { registry }
   const testDir = t.testdir({
     'package.json': JSON.stringify({
       name: 'my-cool-pkg',
@@ -391,7 +410,10 @@ t.test('read registry only from publishConfig', t => {
       flatOptions: {
         json: false,
       },
-      config,
+      config: {
+        ...config,
+        getCredentialsByURI: registryCredentials(t, registry),
+      },
     },
     '../../lib/utils/tar.js': {
       getContents: () => ({
@@ -403,7 +425,7 @@ t.test('read registry only from publishConfig', t => {
     libnpmpublish: {
       publish: (manifest, tarData, opts) => {
         t.match(manifest, { name: 'my-cool-pkg', version: '1.0.0' }, 'gets manifest')
-        t.same(opts.registry, publishConfig.registry, 'publishConfig is passed through')
+        t.same(opts.registry, registry, 'publishConfig is passed through')
       },
     },
   })

